_posts/2019-04-23-autoregressive-flows/experimenting-with-autoregressive-flows.Rmd

```diff
@@ -0,0 +1,645 @@
+
+```{python}
+def _gen_slices(num_blocks, n_in, n_out):
+  slices = []
+  col = 0
+  d_in = n_in // num_blocks
+  d_in
+  d_out = n_out // num_blocks
+  d_out
+  row = d_out 
+  for _ in range(num_blocks):
+    row_slice = slice(row, None)
+    col_slice = slice(col, col + d_in)
+    slices.append([row_slice, col_slice])
+    col += d_in
+    row += d_out
+  return slices
+
+slices = _gen_slices(num_blocks, input_depth, units)
+for [row_slice, col_slice] in slices:
+  mask[row_slice, col_slice] = 1
+
+mask
+```
+
+```
+array([[0., 0., 0.],
+       [1., 0., 0.],
+       [1., 1., 0.],
+       [1., 1., 1.]], dtype=float32)
+```
+
+Again, does this look mirror-inverted? A transpose fixes shape and logic both:
+
+```{python}
+mask = mask.t
+mask
+```
+
+```
+array([[0., 1., 1., 1.],
+       [0., 0., 1., 1.],
+       [0., 0., 0., 1.]], dtype=float32)
+```
+
+Now that we have the mask, we can create the layer (interestingly, as of this writing not (yet?) a `tf.keras` layer):
+
+```{python}
+layer = tf.compat.v1.layers.Dense(
+        units,
+        kernel_initializer=masked_initializer, # 1
+        kernel_constraint=lambda x: mask * x   # 2
+        )
+
+```
+
+Here we see masking going on in two ways. For one, the weight initializer is masked:
+
+```{python}
+kernel_initializer = tf.compat.v1.glorot_normal_initializer()
+
+def masked_initializer(shape, dtype=None, partition_info=None):
+ return mask * kernel_initializer(shape, dtype, partition_info)
+```
+
+And secondly, a kernel constraint makes sure that after optimization, the relative units are zeroed out again:
+
+```{python}
+kernel_constraint=lambda x: mask * x 
+```
+
+Just for fun, let's apply the layer to our toy input:
+
+```{python}
+layer.apply(inputs)
+```
+
+```
+<tf.Tensor: id=30, shape=(2, 4), dtype=float64, numpy=
+array([[ 0.        , -0.7489589 , -0.43329933,  1.42710014],
+       [ 0.        , -2.9958356 , -1.71647246,  1.09258015]])>
+```
+
+Zeroes where expected. And double-checking on the weight matrix...
+
+```{python}
+layer.kernel
+```
+
+```
+<tf.Variable 'dense/kernel:0' shape=(3, 4) dtype=float64, numpy=
+array([[ 0.        , -0.7489589 , -0.42214942, -0.6473454 ],
+       [-0.        ,  0.        , -0.00557496, -0.46692933],
+       [-0.        , -0.        , -0.        ,  1.00276807]])>
+```
+
+Good. Now hopefully after this little deep dive, things have become a bit more concrete. Of course in a bigger model, the autoregressive property has to be conserved between layers as well.
+
+On to the second topic, application of MAF to a real-world dataset.
+
+## Masked Autoregressive Flow 
+
+The MAF paper[@2017arXiv170507057P] applied masked autoregressive flows (as well as single-layer-_MADE_[@GermainGML15] and Real NVP [@DinhSB16]) to a number of datasets, including MNIST, CIFAR-10 and several datasets from the [UCI Machine Learning Repository](http://archive.ics.uci.edu/ml/index.html).
```

Comment:
I think the convention is to write about a paper in the present tense "The MAF paper **_applies_** masked…"